cdap-data-fabric/src/main/java/co/cask/cdap/data2/transaction/AbstractTransactionContext.java

```diff
@@ -0,0 +1,308 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.transaction;
+
+import co.cask.cdap.api.common.Bytes;
+import com.google.common.base.Preconditions;
+import org.apache.tephra.Transaction;
+import org.apache.tephra.TransactionAware;
+import org.apache.tephra.TransactionConflictException;
+import org.apache.tephra.TransactionContext;
+import org.apache.tephra.TransactionFailureException;
+import org.apache.tephra.TransactionSystemClient;
+
+import java.util.Set;
+import java.util.TreeSet;
+import javax.annotation.Nullable;
+
+/**
+ * An abstract implementation of {@link TransactionContext} for governing general transaction lifecycle.
+ * This class provides flexibility on the set of
+ * {@link TransactionAware} to be participated in transactions.
+ *
+ * This abstract class extends from {@link TransactionContext} just for inheriting the type and public methods
```

Comment:
There are couple (~10) JIRAs in Tephra about improving the `TransactionContext`. We can make the decision on how to refactor when designing for those improvements.